### PR TITLE
Fix several bugs found by gvisor test suite

### PR DESCRIPTION
**Pull request by jessehui**
This PR can fix several bugs that are found by running gvisor test suite.

All the fixed gvisor tests are：
bad_test,  futex_test,  poll_test,  preadv2_test,  pwritev2_test,  readv_test,  rtsignal_test,  select_test,  socket_blocking_ip_test.

### CODE REVIEW

File: src/libos/src/poll/do_poll.rs [hunk 1]
```diff
@@ -55,8 +52,12 @@ pub async fn do_poll(poll_fds: &[PollFd], mut timeout: Option<&mut Duration>) ->
         }
 
         // Return if the timeout expires.
-        if let Err(_) = poller.wait_timeout(timeout.as_mut()).await {
-            return Ok(0);
+        if let Err(e) = poller.wait_timeout(timeout.as_mut()).await {
+            if e.errno() != EINTR {
```
Comment (guzongmin) [inline src/libos/src/poll/do_poll.rs:56]: Should check ETIMEDOUT instead of EINTR

File: src/libos/src/fs/file_ops/sendfile.rs [hunk 2]
```diff
@@ -45,8 +45,8 @@ pub async fn do_sendfile(
 
     // If offset is not none, does not modify file offset of in_fd;
     // otherwise the file offset is adjusted to reflect the number of bytes read
-    if offset.is_none() {
-        in_inode_file.seek(SeekFrom::Current(bytes_read as i64))?;
+    if offset.is_none() && bytes_read > 0 {
```
Comment (liqinggd) [inline src/libos/src/fs/file_ops/sendfile.rs:48]: I think this modification is wrong.

In this implementation, if read_bytes is 0 or 1, the offset will not be changed, this does not make sense. If read N bytes, the offset should be added by N, not (N-1).

File: src/libos/crates/async-io/src/util/channel.rs [hunk 3]
```diff
@@ -191,7 +191,11 @@ impl File for Producer {
     // TODO: implement writev
 
     fn poll(&self, mask: Events, poller: Option<&mut Poller>) -> Events {
-        self.this_end().pollee().poll(mask, poller)
+        let mut revents = self.this_end().pollee().poll(mask, poller);
```
Comment (guzongmin) [inline src/libos/crates/async-io/src/util/channel.rs:194]: Please add more comments about the "poll" behavior

File: src/libos/crates/async-io/src/file/file.rs [hunk 4]
```diff
@@ -20,12 +20,17 @@ pub trait File: Debug + Sync + Send {
     }
 
     fn readv(&self, bufs: &mut [&mut [u8]]) -> Result<usize> {
```
Comment (liqinggd) [inline src/libos/crates/async-io/src/file/file.rs:22]: I think the `writev` is also wrong,

**Review by jessehui — COMMENTED** [review]
All done. Thank you for your review.

File: src/libos/src/fs/file_ops/sendfile.rs (hunk 2 above)
Comment (jessehui) [reply src/libos/src/fs/file_ops/sendfile.rs]: This is not a simple offset error but the read length and the write length mismatch for the socket file. Fixed.

File: src/libos/crates/async-io/src/util/channel.rs [hunk 5]
```diff
@@ -300,6 +304,12 @@ impl File for Consumer {
     // TODO: implement readv
 
     fn poll(&self, mask: Events, poller: Option<&mut Poller>) -> Events {
+        if self.peer_end().is_shutdown() {
```
Comment (guzongmin) [inline src/libos/crates/async-io/src/util/channel.rs:307]: Should we check the existing event before return Events::HUP?

**Comment by daan3601** [comment]
> This PR can fix several bugs that are found by running gvisor test suite.
> 
> All the fixed gvisor tests are： bad_test, futex_test, poll_test, preadv2_test, pwritev2_test, readv_test, rtsignal_test, select_test, socket_blocking_ip_test.

Except for the futex_test, all the above tests have passed, and the results of futex_test are as follows:
lQLPDhsG0xwC8knMu80FQbDXlvTdJjRhugHb2s08AGYA_1345_187

**Comment by jessehui** [comment]
> > This PR can fix several bugs that are found by running gvisor test suite.
> > All the fixed gvisor tests are： bad_test, futex_test, poll_test, preadv2_test, pwritev2_test, readv_test, rtsignal_test, select_test, socket_blocking_ip_test.
> 
> Except for the futex_test, all the above tests have passed, and the results of futex_test are as follows: lQLPDhsG0xwC8knMu80FQbDXlvTdJjRhugHb2s08AGYA_1345_187

image

Due to the inaccuracy of the timer, the futex timeout tests can fail sometimes. This error will not be fixed in this PR.

File: src/libos/crates/async-io/src/util/channel.rs (hunk 5 above)
Comment (jessehui) [reply src/libos/crates/async-io/src/util/channel.rs]: Done.

**Review by guzongmin — APPROVED** [review]
LGTM

**Pull request merged**